File: server/src/publish.ts
// Open Cloud place publishing (Phase 4 support).
//
// Publishes a local place file (.rbxl or .rbxlx) as a new version of the configured
// experience, so the test harness lands in the place that run_tests reads. Same
// env-only credentials as cloud.ts, plus the universe-places write scope on the key.

import { readFileSync } from "node:fs";
import { extname } from "node:path";

const BASE = "https://apis.roblox.com/universes/v1";

export interface PublishResult {
  ok: boolean;
  versionNumber?: number;
  error?: string;
}

export interface PublishInput {
  filePath: string;
  versionType?: "Published" | "Saved";
}

export async function publishPlace(input: PublishInput): Promise<PublishResult> {
  const key = process.env.ROBLOX_OPEN_CLOUD_KEY;
  const universe = process.env.ROBLOX_UNIVERSE_ID;
  const place = process.env.ROBLOX_PLACE_ID;
  if (!key || !universe || !place) {
    return { ok: false, error: "Missing env: set ROBLOX_OPEN_CLOUD_KEY, ROBLOX_UNIVERSE_ID, ROBLOX_PLACE_ID." };
  }

  let body: Buffer;
  try {
    body = readFileSync(input.filePath);
  } catch (err) {
    return { ok: false, error: `cannot read ${input.filePath}: ${err instanceof Error ? err.message : String(err)}` };
  }

  // Binary .rbxl uses octet-stream; XML .rbxlx uses application/xml.
  const contentType = extname(input.filePath).toLowerCase() === ".rbxlx" ? "application/xml" : "application/octet-stream";
  const versionType = input.versionType ?? "Published";
  const url = `${BASE}/${universe}/places/${place}/versions?versionType=${versionType}`;
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "x-api-key": key, "Content-Type": contentType },
      body: new Uint8Array(body),
    });
    const text = await res.text();
    if (!res.ok) return { ok: false, error: `publish failed: ${res.status} ${text}`.trim() };
    const data = JSON.parse(text) as { versionNumber?: number };
    return { ok: true, versionNumber: data.versionNumber };
  } catch (err) {
    return { ok: false, error: `publish failed: ${err instanceof Error ? err.message : String(err)}` };
  }
}

File: server/src/publish.rs
// Open Cloud place publishing. Uploads a local .rbxl/.rbxlx as a new version of the
// configured experience (needs the universe-places write scope). Same env-only key.
// Binary places use octet-stream; XML places use application/xml.

use std::path::Path;

use reqwest::Client;
use serde_json::Value;

use crate::env;
use crate::httpx;

const BASE: &str = "https://apis.roblox.com/universes/v1";

pub type PublishResult = Result<Value, String>;

pub async fn publish_place(
    http: &Client,
    file_path: &str,
    version_type: Option<&str>,
) -> PublishResult {
    let key = env::var("ROBLOX_OPEN_CLOUD_KEY").ok_or("Missing env: set ROBLOX_OPEN_CLOUD_KEY.")?;
    let universe = env::var("ROBLOX_UNIVERSE_ID").ok_or("Missing env: set ROBLOX_UNIVERSE_ID.")?;
    let place = env::var("ROBLOX_PLACE_ID").ok_or("Missing env: set ROBLOX_PLACE_ID.")?;

    let bytes = std::fs::read(file_path).map_err(|e| format!("cannot read {file_path}: {e}"))?;
    let is_xml = Path::new(file_path)
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.eq_ignore_ascii_case("rbxlx"))
        .unwrap_or(false);
    let content_type = if is_xml {
        "application/xml"
    } else {
        "application/octet-stream"
    };
    let version = version_type.unwrap_or("Published");
    let url = format!("{BASE}/{universe}/places/{place}/versions");

    let res = httpx::send_retrying(|| {
        http.post(&url)
            .header("x-api-key", &key)
            .header("content-type", content_type)
            .query(&[("versionType", version)])
            .body(bytes.clone())
    })
    .await
    .map_err(|e| format!("publish failed: {e}"))?;

    res.json::<Value>().await.map_err(|e| e.to_string())
}
